### ISSUE TITLE
cram timing issue

### ISSUE THREAD

**Issue by elyisgreat** [alda-lang/alda#252]
Consider the following code:

```
(tempo! 60)

piano:
V1: o3 e b e b
V2: {f+14~/a+14~ a+14~/b14~ f+1~/a+1~}2 g+4/b g+/b
```

This plays normally, with the correct timing given the CRAM ratios. However, this code

```
(tempo! 60)

piano:
V1: o3 e b e b
V2: {f+14~/a+~ a+~/b~ f+1~/a+~}2 g+4/b g+/b
```

which *should* play identically to the one before, plays the CRAM for slightly longer and causes a time delay going forward.

**Comment by daveyarwood** [comment]
Wow, this is a weird bug. In the second example, the duration of the cram itself is taking up slightly longer than a half note, despite the duration being a half note. Good find!

I'll take a look when I can, unless someone else can dig in.

**Comment by daveyarwood** [comment]
I found the cause -- first, a little background on how crams work under the hood:

- Crams either "speed up" or "slow down" the notes inside by multiplying their durations by a scale factor. A scale factor of 1 means playing at 1x, i.e. each note has its normal length according to the current tempo. Crams have the ability to make the notes play "faster" by decreasing the scale factor, e.g. a scale factor of 0.5x means each note is only half as long as it would normally be, so it sounds like the notes are playing twice as fast. Going the other way, a scale factor of 2x means each note is twice as long, so it sounds like the notes are playing twice as slow.

- The scale factor is calculated based on two numbers: the "inner duration" and the "outer duration."
  - The outer duration is the number on the end of the cram expression (if it exists; if it doesn't, then the outer duration is whatever the current default duration is for notes outside of the cram expression); in the case of your example, the outer duration is `2`.
  - The inner duration is the total duration of all of the notes inside of the cram expression, taking into account any overlapping notes like you get with chords.

The bug here has to do with the way the total duration of a chord is calculated for the inner duration of the cram. Currently, [we're just ignoring any notes in the chord that don't have an explicit duration](https://github.com/alda-lang/alda/blob/f321902d5b705101523644570240061b9ad0fc30/server/src/alda/lisp/model/duration.clj#L27-L36), so if _none_ of the notes in the chord have an explicit duration (as in your second example), then the duration of the chord is calculated as 0, so the chord is basically skipped. This throws the time scale calculation off because the inner duration is one note smaller than it should be, resulting in the timing skew you're hearing.

Note that this fixes your example:

```
(tempo! 60)

piano:
V1: o3 e b e b
V2: {f+14~/a+~ a+14~/b~ f+1~/a+~}2 g+4/b g+/b
```

The only difference is I changed `a+~/b~` to `a+14~/b~`. As long as at least one of the notes in each chord has an explicit duration, then the cram timing will work the way it should.

When I have a minute, I'll see if I can get it working so that it uses the implicit duration from previous notes/chords inside the cram, so that notes/chords inside the cram are never skipped when calculating the inner duration.

**Comment by daveyarwood** [comment]
By the way, excellent use of "14th notes" inside a half note cram. I love it! 😄

**Comment by elyisgreat** [comment]
yep. I see it as doing ratio calculations (which I think is the idea), so {d14 d c1}2 would be "play the d notes 14 times faster than the c, within the span of a half note." thus {d28 d c2}2 and {d7 d c0.5}2 would be equivalent expressions. Interesting how decimals are allowed in crams but not in regular notes. Why is this?

The reason I used 14, by the way, is because I wanted the last note to play 7 times longer than the first two (simulating a mordent). That way the two notes would take up 1/8 of the cram and the final note would take up 7/8 of the cram.

**Comment by daveyarwood** [comment]
> yep. I see it as doing ratio calculations (which I think is the idea), so {d14 d c1}2 would be "play the d notes 14 times faster than the c, within the span of a half note." thus {d28 d c2}2 and {d7 d c0.5}2 would be equivalent expressions.

Exactly!

> Interesting how decimals are allowed in crams but not in regular notes. Why is this?

They're actually not. `c0.5` results in an "invalid syntax" error both inside and outside of a cram.

> The reason I used 14, by the way, is because I wanted the last note to play 7 times longer than the first two (simulating a mordent). That way the two notes would take up 1/8 of the cram and the final note would take up 7/8 of the cram.

Yep, makes sense!

**Comment by daveyarwood** [comment]
Fixed in [1.0.0-rc29](https://github.com/alda-lang/alda/releases/tag/1.0.0-rc29).
